feat: Add APIContext object to Auth Config

JordanSekky opened the pull request:
Adds an option to the Auth Config file where users can define a function which accepts the Astro global object (for Astro pages/components) or the APIContext object (for API Routes).

This allows Cloudflare D1 users to use the environment provided by the cloudflare bindings, rather the the vite environment variables.

Adds getSessionByContext(), which allow fetching the session when using a lazy auth config.

Fixes nowaythatworked#60, nowaythatworked#52, nowaythatworked#50.

nowaythatworked commented on server.ts:
```diff
@@ -86,17 +90,23 @@ export function AstroAuth(options = authConfig) {
-	const handler = AstroAuthHandler(prefix, authOptions)
 	return {
 		async GET(context: APIContext) {
+			const config = isUserConfigLazy(options) ? await options.config(context) : options
+			config.secret ??= AUTH_SECRET
+			config.trustHost ??= !!(AUTH_TRUST_HOST ?? VERCEL ?? NODE_ENV !== 'production')
+
+			const { prefix = '/api/auth', ...authOptions } = config
+			const handler = AstroAuthHandler(prefix, authOptions)
 			return await handler(context)
 		},
 		async POST(context: APIContext) {
+			const config = isUserConfigLazy(options) ? await options.config(context) : options
+			config.secret ??= AUTH_SECRET
+			config.trustHost ??= !!(AUTH_TRUST_HOST ?? VERCEL ?? NODE_ENV !== 'production')
+
+			const { prefix = '/api/auth', ...authOptions } = config
+			const handler = AstroAuthHandler(prefix, authOptions)
```
would be nice to have this a bit cleaner, not so repetitive.

vtyagunov replied:
Hi @nowaythatworked @JordanSekky 

I think this can be moved to the `AstroAuthHandler` implementation.
Resolving config with `await options.config(context)` is also could be a part of that handler.

```ts
function AstroAuthHandler(lazyConfig = authConfig) {
  return async (context: APIContext) => {
    const config = isUserConfigLazy(lazyConfig) ? await lazyConfig.config(context) : lazyConfig 
    config.secret ??= AUTH_SECRET
    config.trustHost ??= !!(AUTH_TRUST_HOST ?? VERCEL ?? NODE_ENV !== 'production')
    
    const { prefix = '/api/auth', ...options } = config
    
    // The rest remains the same.
  }
}


export function AstroAuth(options = authConfig) {
	const handler = AstroAuthHandler(options)
	return {
		async GET(context: APIContext) {
			return await handler(context)
		},
		async POST(context: APIContext) {
			return await handler(context)
		},
	}
}
 ```

nowaythatworked commented on server.ts:
```diff
@@ -108,10 +118,14 @@ export function AstroAuth(options = authConfig) {
  * @returns The current session, or `null` if there is no session.
  */
 export async function getSession(req: Request, options = authConfig): Promise<Session | null> {
+	if (isUserConfigLazy(options)) {
+		throw new Error(
+			'User Auth Configuration is Lazy. Fetch the session using getSessionByContext().'
+		)
+	}
```
Since you're already checking for lazy config, we could have the functionality of `getSessionByContext` directly in here, right?
The `req` argument would need to be replaced with context (AstroGlobal, etc) which would be a breaking change, but I think that would be fine? Let me know what you think.

nowaythatworked reviewed (commented):
Thanks for putting in the work. Looks great already, left some comments with my thoughts :)

JordanSekky committed:
feat: Add APIContext object to Auth Config

Adds an option to the Auth Config file where users can define
a function which accept the Astro global object (for Astro
pages/components) or the APIContext object (for API Routes).

This allows Cloudflare D1 users to use the environment provided by
the cloudflare bindings, rather the the vite environment variables.

Adds getSessionByContext(), which allows fetching the session
when using a lazy auth config.

Fixes nowaythatworked#60, nowaythatworked#52, nowaythatworked#50.

vtyagunov commented on src/config.ts:
```diff
@@ -43,3 +45,9 @@ export const defineConfig = (config: FullAuthConfig) => {
 	return config
 }
+
+export type UserAuthConfig =
+	| {
+			config: (ctx: APIContext | AstroGlobal | ActionAPIContext) => FullAuthConfig | Promise<FullAuthConfig>
+	  }
+	| FullAuthConfig
```
What do you guys think about this?

```suggestion
export type LazyAuthConfig = (ctx: APIContext | AstroGlobal | ActionAPIContext) => FullAuthConfig | Promise<FullAuthConfig>

export type UserAuthConfig =
	| LazyAuthConfig
	| FullAuthConfig
```
